app/src/main/java/org/fossasia/susi/ai/login/LoginPresenter.kt

```diff
@@ -165,4 +170,56 @@ class LoginPresenter(loginActivity: LoginActivity) : ILoginPresenter, ILoginMode
     override fun onDetach() {
         loginView = null
     }
+
+    override fun requestPassword(email: String, url: String, isPersonalServerChecked: Boolean) {
+        if (email.isEmpty()) {
+            loginView?.invalidCredentials(true, Constant.EMAIL)
+            return
+        }
+
+        if (!CredentialHelper.isEmailValid(email)) {
+            loginView?.invalidCredentials(false, Constant.EMAIL)
+            return
+        }
+
+        if (isPersonalServerChecked) {
+            if (url.isEmpty()) {
+                loginView?.invalidCredentials(true, Constant.INPUT_URL)
+                return
+            }
+            if (CredentialHelper.isURLValid(url)) {
+                if (CredentialHelper.getValidURL(url) != null) {
+                    utilModel.setServer(false)
+                    utilModel.setCustomURL(CredentialHelper.getValidURL(url) as String)
+                } else {
+                    loginView?.invalidCredentials(false, Constant.INPUT_URL)
+                    return
+                }
+            } else {
+                loginView?.invalidCredentials(false, Constant.INPUT_URL)
+                return
+            }
+        } else {
+            utilModel.setServer(true)
+        }
+        this.email = email
+        loginView?.showForgotPasswordProgress(true)
+        forgotPasswordModel.requestPassword(email.trim({ it <= ' ' }), this)
+    }
+
+    override fun onForgotPasswordModelSuccess(response: Response<ForgotPasswordResponse>) {
+        loginView?.showForgotPasswordProgress(false)
+        if (response.isSuccessful && response.body() != null) {
+            loginView?.resetPasswordSuccess(utilModel.getString(R.string.forgot_password_mail_sent), response.body().message)
+        } else if (response.code() == 422) {
+            loginView?.resetPasswordFailure(utilModel.getString(R.string.email_invalid_title), utilModel.getString(R.string.email_invalid), utilModel.getString(R.string.retry), Color.RED)
+        } else {
+            loginView?.resetPasswordFailure("${response.code()} " + utilModel.getString(R.string.error), response.message(), utilModel.getString(R.string.ok), Color.BLUE)
+        }
+    }
+
+    override fun cancelSignup() {
+        forgotPasswordModel.cancelSignup()
+    }
+
```

Comment:
The project is using MVP but all contributors seem to just put everything in the views. This is MVC guys, which is untestable even when using isolated components. And this is not even isolated. This is creating God Activities. Please refactor your code to follow SOLID and think about design as well. The goal of GSoC is to let you experience and become much better developers who can code beyond their comfort zone.